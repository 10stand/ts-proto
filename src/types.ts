import { google } from '../build/pbjs';
import { TypeName, TypeNames } from 'ts-poet';
import { mapMessageType } from './main';
import FieldDescriptorProto = google.protobuf.FieldDescriptorProto;

/** Based on https://github.com/dcodeIO/protobuf.js/blob/master/src/types.js#L37. */
export function basicWireType(type: FieldDescriptorProto.Type): number {
  switch (type) {
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
      return 1;
    case FieldDescriptorProto.Type.TYPE_FLOAT:
      return 5;
    case FieldDescriptorProto.Type.TYPE_INT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_UINT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_SINT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_FIXED32:
      return 5;
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return 5;
    case FieldDescriptorProto.Type.TYPE_INT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_UINT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_SINT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_FIXED64:
      return 1;
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      return 1;
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return 0;
    case FieldDescriptorProto.Type.TYPE_STRING:
      return 2;
    case FieldDescriptorProto.Type.TYPE_BYTES:
      return 2;
    default:
      throw new Error('Invalid type ' + type);
  }
}

/** Returns the type name without any repeated/required/etc. labels. */
export function basicTypeName(field: FieldDescriptorProto): TypeName {
  switch (field.type) {
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
    case FieldDescriptorProto.Type.TYPE_FLOAT:
    case FieldDescriptorProto.Type.TYPE_INT32:
    case FieldDescriptorProto.Type.TYPE_UINT32:
    case FieldDescriptorProto.Type.TYPE_SINT32:
    case FieldDescriptorProto.Type.TYPE_FIXED32:
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return TypeNames.NUMBER;
    case FieldDescriptorProto.Type.TYPE_INT64:
    case FieldDescriptorProto.Type.TYPE_UINT64:
    case FieldDescriptorProto.Type.TYPE_SINT64:
    case FieldDescriptorProto.Type.TYPE_FIXED64:
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      // type = config.forceLong ? "Long" : config.forceNumber ? "number" : "number|Long";
      return TypeNames.NUMBER;
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return TypeNames.BOOLEAN;
    case FieldDescriptorProto.Type.TYPE_STRING:
      return TypeNames.STRING;
    case FieldDescriptorProto.Type.TYPE_BYTES:
      return TypeNames.anyType('Uint8Array');
    case FieldDescriptorProto.Type.TYPE_MESSAGE:
    case FieldDescriptorProto.Type.TYPE_ENUM:
      return mapMessageType(field.typeName);
    default:
      return TypeNames.anyType(field.typeName);
  }
}

/** Returns the Reader method for the primitive's read/write call. */
export function toReaderCall(field: FieldDescriptorProto): string {
  switch (field.type) {
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
      return 'double';
    case FieldDescriptorProto.Type.TYPE_FLOAT:
      return 'float';
    case FieldDescriptorProto.Type.TYPE_INT32:
    case FieldDescriptorProto.Type.TYPE_ENUM:
      return 'int32';
    case FieldDescriptorProto.Type.TYPE_UINT32:
      return 'uint32';
    case FieldDescriptorProto.Type.TYPE_SINT32:
      return 'sint32';
    case FieldDescriptorProto.Type.TYPE_FIXED32:
      return 'fixed32';
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return 'sfixed32';
    case FieldDescriptorProto.Type.TYPE_INT64:
      return 'int64';
    case FieldDescriptorProto.Type.TYPE_UINT64:
      return 'uint64';
    case FieldDescriptorProto.Type.TYPE_SINT64:
      return 'sint64';
    case FieldDescriptorProto.Type.TYPE_FIXED64:
      return 'fixed64';
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      return 'sfixed64';
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return 'bool';
    case FieldDescriptorProto.Type.TYPE_STRING:
      return 'string';
    case FieldDescriptorProto.Type.TYPE_BYTES:
      return 'bytes';
    default:
      throw new Error(`Not a primitive field ${field}`);
  }
}

export function packedType(type: FieldDescriptorProto.Type): number | undefined {
  switch (type) {
    case FieldDescriptorProto.Type.TYPE_DOUBLE:
      return 1;
    case FieldDescriptorProto.Type.TYPE_FLOAT:
      return 5;
    case FieldDescriptorProto.Type.TYPE_INT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_UINT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_SINT32:
      return 0;
    case FieldDescriptorProto.Type.TYPE_FIXED32:
      return 5;
    case FieldDescriptorProto.Type.TYPE_SFIXED32:
      return 5;
    case FieldDescriptorProto.Type.TYPE_INT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_UINT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_SINT64:
      return 0;
    case FieldDescriptorProto.Type.TYPE_FIXED64:
      return 1;
    case FieldDescriptorProto.Type.TYPE_SFIXED64:
      return 1;
    case FieldDescriptorProto.Type.TYPE_BOOL:
      return 0;
    default:
      return undefined;
  }
}
