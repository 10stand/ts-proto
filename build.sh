#!/usr/bin/env bash

./pbjs.sh
./integration/pbjs.sh
(cd ./integration || exit; ./codegen.sh)

yarn test