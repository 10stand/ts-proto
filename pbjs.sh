#!/bin/bash

# Generates objects with pbjs so we can bootstrap.

mkdir -p build
yarn pbjs --force-message -t static-module -o build/pbjs.js \
  google/protobuf/descriptor.proto \
  google/protobuf/compiler/plugin.proto \
  simple.proto
yarn pbts --no-comments -o build/pbjs.d.ts build/pbjs.js

