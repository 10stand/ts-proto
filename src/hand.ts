import { ts_proto_tests } from "../build/pbjs";
import ISimple = ts_proto_tests.ISimple;
import PbSimple = ts_proto_tests.Simple;
import { Reader } from "protobufjs";

interface Simple {
  name: string;
}

function decode(reader: Reader, length?: number): Simple {
  let end = length === undefined ? reader.len : reader.pos + length;
  const message = {} as Simple;
  while (reader.pos < end) {
    const tag = reader.uint32();
    switch (tag >>> 3) {
      case 1:
        message.name = reader.string();
        break;
      default:
        reader.skipType(tag & 7);
        break;
    }
  }
  return message;
}

function main() {
  const s: ISimple = { name: "asdf" };
  const ss = new PbSimple(s);
  const writer = PbSimple.encode(ss);
  const bytes = writer.finish();
  console.log(bytes);

  const s2 = PbSimple.decode(bytes);
  console.log(s2.name);

  const s3 = decode(Reader.create(bytes));
  console.log(s3.name);

}


main();