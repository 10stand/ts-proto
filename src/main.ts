import {
  ClassSpec,
  CodeBlock,
  EnumSpec,
  FileSpec,
  FunctionSpec,
  InterfaceSpec,
  Member,
  Modifier,
  PropertySpec,
  TypeName,
  TypeNames
} from 'ts-poet';
import { google } from '../build/pbjs';
import {
  basicLongWireType,
  basicTypeName,
  basicWireType,
  defaultValue,
  isMessage,
  isPrimitive,
  isRepeated,
  isValueType,
  isWithinOneOf,
  messageToTypeName,
  packedType,
  toReaderCall,
  toTypeName,
  TypeMap
} from './types';
import { asSequence } from 'sequency';
import DescriptorProto = google.protobuf.DescriptorProto;
import FieldDescriptorProto = google.protobuf.FieldDescriptorProto;
import FileDescriptorProto = google.protobuf.FileDescriptorProto;
import EnumDescriptorProto = google.protobuf.EnumDescriptorProto;
import ServiceDescriptorProto = google.protobuf.ServiceDescriptorProto;
import MethodDescriptorProto = google.protobuf.MethodDescriptorProto;

export function generateFile(typeMap: TypeMap, fileDesc: FileDescriptorProto): FileSpec {
  const moduleName = fileDesc.name.replace('.proto', '').replace(/\//g, '_');
  let file = FileSpec.create(moduleName);

  // first make all the type declarations
  visit(
    fileDesc,
    (fullName, message) => {
      file = file.addInterface(generateInterfaceDeclaration(typeMap, fullName, message));
    },
    (fullName, enumDesc) => {
      file = file.addEnum(generateEnum(fullName, enumDesc));
    }
  );

  // then add the encoder/decoder/base instance
  visit(fileDesc, (fullName, message) => {
    file = file.addProperty(generateBaseInstance(fullName, message));
    let staticMethods = CodeBlock.empty()
      .add('export const %L = ', fullName)
      .beginHash()
      .addHashEntry('encode', generateEncode(typeMap, fullName, message))
      .addHashEntry('decode', generateDecode(typeMap, fullName, message))
      .endHash()
      .add(';')
      .newLine();
    file = file.addCode(staticMethods);
  });

  visitServices(fileDesc, serviceDesc => {
    file = file.addInterface(generateService(typeMap, serviceDesc));
    file = file.addClass(generateServiceClientImpl(typeMap, serviceDesc));
  });
  file = file.addInterface(generateRpcType());

  file = addLongUtilityMethod(file);

  return file;
}

function addLongUtilityMethod(file: FileSpec): FileSpec {
  return file.addFunction(
    FunctionSpec.create('longToNumber')
      .addParameter('long', 'Long*long')
      .addCodeBlock(
        CodeBlock.empty()
          .beginControlFlow('if (long.gt(Number.MAX_VALUE))')
          .addStatement('throw new Error("Value is larger than Number.MAX_VALUE");')
          .endControlFlow()
          .addStatement('return long.toNumber()')
      )
  );
}

function generateEnum(fullName: string, enumDesc: EnumDescriptorProto): EnumSpec {
  let spec = EnumSpec.create(fullName).addModifiers(Modifier.EXPORT);
  for (const valueDesc of enumDesc.value) {
    spec = spec.addConstant(valueDesc.name, valueDesc.number.toString());
  }
  return spec;
}

// Create the interface with properties
function generateInterfaceDeclaration(typeMap: TypeMap, fullName: string, messageDesc: DescriptorProto) {
  let message = InterfaceSpec.create(fullName).addModifiers(Modifier.EXPORT);
  for (const fieldDesc of messageDesc.field) {
    message = message.addProperty(PropertySpec.create(snakeToCamel(fieldDesc.name), toTypeName(typeMap, fieldDesc)));
  }
  return message;
}

function generateBaseInstance(fullName: string, messageDesc: DescriptorProto) {
  // Create a 'base' instance with default values for decode to use as a prototype
  let baseMessage = PropertySpec.create('base' + fullName, TypeNames.anyType('object')).addModifiers(Modifier.CONST);
  let initialValue = CodeBlock.empty().beginHash();
  asSequence(messageDesc.field)
    .filterNot(isWithinOneOf)
    .forEach(field => {
      initialValue = initialValue.addHashEntry(snakeToCamel(field.name), defaultValue(field.type));
    });
  return baseMessage.initializerBlock(initialValue.endHash());
}

type MessageVisitor = (fullName: string, desc: DescriptorProto) => void;
type EnumVisitor = (fullName: string, desc: EnumDescriptorProto) => void;
export function visit(
  proto: FileDescriptorProto | DescriptorProto,
  messageFn: MessageVisitor,
  enumFn: EnumVisitor = () => {},
  prefix: string = ''
): void {
  for (const enumDesc of proto.enumType) {
    const fullName = prefix + snakeToCamel(enumDesc.name);
    enumFn(fullName, enumDesc);
  }
  const messages = proto instanceof FileDescriptorProto ? proto.messageType : proto.nestedType;
  for (const message of messages) {
    const fullName = prefix + snakeToCamel(message.name);
    messageFn(fullName, message);
    visit(message, messageFn, enumFn, fullName + '_');
  }
}

function visitServices(proto: FileDescriptorProto, serviceFn: (desc: ServiceDescriptorProto) => void): void {
  for (const serviceDesc of proto.service) {
    serviceFn(serviceDesc);
  }
}

/** Creates a function to decode a message by loop overing the tags. */
function generateDecode(typeMap: TypeMap, fullName: string, messageDesc: DescriptorProto): FunctionSpec {
  // create the basic function declaration
  let func = FunctionSpec.create('decode' + fullName)
    .addParameter('reader', 'Reader@protobufjs/minimal')
    .addParameter('length?', 'number')
    .returns(fullName);

  // add the initial end/message
  func = func
    .addStatement('let end = length === undefined ? reader.len : reader.pos + length')
    .addStatement('const message = Object.create(base%L) as %L', fullName, fullName);

  // initialize all lists
  messageDesc.field.filter(isRepeated).forEach(field => {
    func = func.addStatement('message.%L = []', snakeToCamel(field.name));
  });

  // start the tag loop
  func = func
    .beginControlFlow('while (reader.pos < end)')
    .addStatement('const tag = reader.uint32()')
    .beginControlFlow('switch (tag >>> 3)');

  // add a case for each incoming field
  messageDesc.field.forEach(field => {
    const fieldName = snakeToCamel(field.name);
    func = func.addCode('case %L:%>\n', field.number);

    // get a generic 'reader.doSomething' bit that is specific to the basic type
    let readSnippet: CodeBlock;
    if (isPrimitive(field)) {
      readSnippet = CodeBlock.of('reader.%L()', toReaderCall(field));
      if (basicLongWireType(field.type) !== undefined) {
        readSnippet = CodeBlock.of('longToNumber(%L as Long)', readSnippet);
      }
    } else if (isValueType(field)) {
      readSnippet = CodeBlock.of('%T.decode(reader, reader.uint32()).value', basicTypeName(typeMap, field, true));
    } else if (isMessage(field)) {
      readSnippet = CodeBlock.of('%T.decode(reader, reader.uint32())', basicTypeName(typeMap, field));
    } else {
      throw new Error(`Unhandled field ${field}`);
    }

    // and then use the snippet to handle repeated fields if necessary
    if (isRepeated(field)) {
      if (packedType(field.type) === undefined) {
        func = func.addStatement(`message.%L.push(%L)`, fieldName, readSnippet);
      } else {
        func = func
          .beginControlFlow('if ((tag & 7) === 2)')
          .addStatement('const end2 = reader.uint32() + reader.pos')
          .beginControlFlow('while (reader.pos < end2)')
          .addStatement(`message.%L.push(%L)`, fieldName, readSnippet)
          .endControlFlow()
          .nextControlFlow('else')
          .addStatement(`message.%L.push(%L)`, fieldName, readSnippet)
          .endControlFlow();
      }
    } else {
      func = func.addStatement(`message.%L = %L`, fieldName, readSnippet);
    }
    func = func.addStatement('break%<');
  });
  func = func
    .addCode('default:%>\n')
    .addStatement('reader.skipType(tag & 7)')
    .addStatement('break%<');
  // and then wrap up the switch/while/return
  func = func
    .endControlFlow()
    .endControlFlow()
    .addStatement('return message');
  return func;
}

/** Creates a function to encode a message by loop overing the tags. */
function generateEncode(typeMap: TypeMap, fullName: string, messageDesc: DescriptorProto): FunctionSpec {
  // create the basic function declaration
  let func = FunctionSpec.create('encode' + fullName)
    .addParameter('message', fullName)
    .addParameter('writer', 'Writer@protobufjs/minimal', { defaultValueField: CodeBlock.of('new Writer()') })
    .returns('Writer@protobufjs/minimal');
  // then add a case for each field
  messageDesc.field.forEach(field => {
    const fieldName = snakeToCamel(field.name);

    // get a generic writer.doSomething based on the basic type
    let writeSnippet: (place: string) => CodeBlock;
    if (isPrimitive(field)) {
      const tag = ((field.number << 3) | basicWireType(field.type)) >>> 0;
      writeSnippet = place => CodeBlock.of('writer.uint32(%L).%L(%L)', tag, toReaderCall(field), place);
    } else if (isValueType(field)) {
      const tag = ((field.number << 3) | 2) >>> 0;
      writeSnippet = place =>
        CodeBlock.of(
          '%T.encode({ value: %L! }, writer.uint32(%L).fork()).ldelim()',
          basicTypeName(typeMap, field, true),
          place,
          tag
        );
    } else if (isMessage(field)) {
      const tag = ((field.number << 3) | 2) >>> 0;
      writeSnippet = place =>
        CodeBlock.of('%T.encode(%L, writer.uint32(%L).fork()).ldelim()', basicTypeName(typeMap, field), place, tag);
    } else {
      throw new Error(`Unhandled field ${field}`);
    }

    if (isRepeated(field)) {
      if (packedType(field.type) === undefined) {
        func = func
          .beginControlFlow('for (const v of message.%L)', fieldName)
          .addStatement('%L', writeSnippet('v!'))
          .endControlFlow();
      } else {
        const tag = ((field.number << 3) | 2) >>> 0;
        func = func
          .addStatement('writer.uint32(%L).fork()', tag)
          .beginControlFlow('for (const v of message.%L)', fieldName)
          .addStatement('writer.%L(v)', toReaderCall(field))
          .endControlFlow()
          .addStatement('writer.ldelim()');
      }
    } else if (isWithinOneOf(field) || isMessage(field)) {
      func = func
        .beginControlFlow(
          'if (message.%L !== undefined && message.%L !== %L)',
          fieldName,
          fieldName,
          defaultValue(field.type)
        )
        .addStatement('%L', writeSnippet(`message.${fieldName}`))
        .endControlFlow();
    } else {
      func = func.addStatement('%L', writeSnippet(`message.${fieldName}`));
    }
  });
  return func.addStatement('return writer');
}

function generateService(typeMap: TypeMap, serviceDesc: ServiceDescriptorProto): InterfaceSpec {
  let service = InterfaceSpec.create(serviceDesc.name).addModifiers(Modifier.EXPORT);
  for (const methodDesc of serviceDesc.method) {
    service = service.addFunction(
      FunctionSpec.create(methodDesc.name)
        .addModifiers(Modifier.ABSTRACT)
        .addParameter('request', requestType(typeMap, methodDesc))
        .returns(responsePromise(typeMap, methodDesc))
    );
  }
  return service;
}

function generateServiceClientImpl(typeMap: TypeMap, serviceDesc: ServiceDescriptorProto): ClassSpec {
  let client = ClassSpec.create(`${serviceDesc.name}ClientImpl`).addModifiers(Modifier.EXPORT);
  client = client.addFunction(
    FunctionSpec.createConstructor()
      .addParameter('rpc', 'Rpc')
      .addStatement('this.rpc = rpc')
  );
  client = client.addProperty('rpc', 'Rpc', { modifiers: [Modifier.PRIVATE, Modifier.READONLY] });
  for (const methodDesc of serviceDesc.method) {
    client = client.addFunction(
      FunctionSpec.create(methodDesc.name)
        .addParameter('request', requestType(typeMap, methodDesc))
        .addStatement('const data = %L.encode(request).finish()', requestType(typeMap, methodDesc))
        .addStatement('const promise = this.rpc.request(%S, %S, %L)', serviceDesc.name, methodDesc.name, 'data')
        .addStatement('return promise.then(data => %L.decode(new Reader(data)))', responseType(typeMap, methodDesc))
        .returns(responsePromise(typeMap, methodDesc))
    );
  }
  return client;
}

function generateRpcType(): InterfaceSpec {
  const data = TypeNames.anyType('Uint8Array');
  return InterfaceSpec.create('Rpc')
    .addModifiers(Modifier.EXPORT)
    .addFunction(
      FunctionSpec.create('request')
        .addModifiers(Modifier.ABSTRACT)
        .addParameter('service', TypeNames.STRING)
        .addParameter('method', TypeNames.STRING)
        .addParameter('data', data)
        .returns(TypeNames.PROMISE.param(data))
    );
}

function requestType(typeMap: TypeMap, methodDesc: MethodDescriptorProto): TypeName {
  return messageToTypeName(typeMap, methodDesc.inputType);
}

function responseType(typeMap: TypeMap, methodDesc: MethodDescriptorProto): TypeName {
  return messageToTypeName(typeMap, methodDesc.outputType);
}

function responsePromise(typeMap: TypeMap, methodDesc: MethodDescriptorProto): TypeName {
  return TypeNames.PROMISE.param(responseType(typeMap, methodDesc));
}

function generateOneOfProperty(typeMap: TypeMap, name: string, fields: FieldDescriptorProto[]): PropertySpec {
  const adtType = TypeNames.unionType(
    ...fields.map(f => {
      const kind = new Member('field', TypeNames.anyType(`'${f.name}'`), false);
      const value = new Member('value', toTypeName(typeMap, f), false);
      return TypeNames.anonymousType(kind, value);
    })
  );
  return PropertySpec.create(snakeToCamel(name), adtType);
}

function snakeToCamel(s: string): string {
  return s.replace(/(\_\w)/g, m => m[1].toUpperCase());
}
