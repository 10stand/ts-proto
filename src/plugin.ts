import { readToBuffer } from "./utils";
import { google } from "../build/pbjs";
import CodeGeneratorRequest = google.protobuf.compiler.CodeGeneratorRequest;
import { generateFile } from "./main";

// this would be the plugin called by the protoc compiler
async function main() {
  const stdin = await readToBuffer(process.stdin);
  // const json = JSON.parse(stdin.toString());
  // const request = CodeGeneratorRequest.fromObject(json);
  const request = CodeGeneratorRequest.decode(stdin);
  for (let file of request.protoFile) {
    generateFile(file);
  }
  // TODO write to stdout
}


