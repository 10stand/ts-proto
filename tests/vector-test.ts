import { google } from "../build/pbjs";
import CodeGeneratorRequest = google.protobuf.compiler.CodeGeneratorRequest;
import { readFileSync } from "fs";
import { generateFile } from "../src/main";
import { StringBuffer } from "ts-poet/build/StringBuffer";

describe('vector-tile', () => {
  it('works', () => {
    const stdin = readFileSync('./vector_tile.bin');
    const request = CodeGeneratorRequest.decode(stdin);
    for (let file of request.protoFile) {
      const spec = generateFile(file);
      const out = new StringBuffer();
      spec.emit(out);
      expect(out.toString()).toMatchSnapshot();
    }
  });
});
