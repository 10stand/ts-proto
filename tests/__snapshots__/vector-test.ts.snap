// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vector-tile works 1`] = `
"import {Reader, Writer} from 'protobufjs';


export function decodeValue(reader: Reader, length?: number): Value {
  let end = length === undefined ? reader.len : reader.pos + length;
  const message = {} as Value;
  while (reader.pos < end) {
    const tag = reader.uint32();
    switch (tag >>> 3) {
      case 1:
      message.string_value = reader.string();
      break;
      case 2:
      message.float_value = reader.float();
      break;
      case 3:
      message.double_value = reader.double();
      break;
      case 4:
      message.int_value = reader.int64();
      break;
      case 5:
      message.uint_value = reader.uint64();
      break;
      case 6:
      message.sint_value = reader.sint64();
      break;
      case 7:
      message.bool_value = reader.bool();
      break;
      default:
      reader.skipType(tag & 7);
      break;
    }
  }
  return message;
}

export function encodeValue(message: Value, writer: Writer = new Writer()): Writer {
  writer.uint32(10).string(message.string_value);
  writer.uint32(21).float(message.float_value);
  writer.uint32(25).double(message.double_value);
  writer.uint32(32).int64(message.int_value);
  writer.uint32(40).uint64(message.uint_value);
  writer.uint32(48).sint64(message.sint_value);
  writer.uint32(56).bool(message.bool_value);
  return writer;
}

export interface Tile_Value {

  string_value: string;

  float_value: number;

  double_value: number;

  int_value: number;

  uint_value: number;

  sint_value: number;

  bool_value: boolean;

}

export function decodeFeature(reader: Reader, length?: number): Feature {
  let end = length === undefined ? reader.len : reader.pos + length;
  const message = {} as Feature;
  while (reader.pos < end) {
    const tag = reader.uint32();
    switch (tag >>> 3) {
      case 1:
      message.id = reader.uint64();
      break;
      case 2:
      message.tags = reader.uint32();
      break;
      case 3:
      message.type = reader.int32();
      break;
      case 4:
      message.geometry = reader.uint32();
      break;
      default:
      reader.skipType(tag & 7);
      break;
    }
  }
  return message;
}

export function encodeFeature(message: Feature, writer: Writer = new Writer()): Writer {
  writer.uint32(8).uint64(message.id);
  writer.uint32(16).uint32(message.tags);
  writer.uint32(24).int32(message.type);
  writer.uint32(32).uint32(message.geometry);
  return writer;
}

export interface Tile_Feature {

  id: number;

  tags: Array<number>;

  type: Tile_GeomType;

  geometry: Array<number>;

}

export function decodeLayer(reader: Reader, length?: number): Layer {
  let end = length === undefined ? reader.len : reader.pos + length;
  const message = {} as Layer;
  while (reader.pos < end) {
    const tag = reader.uint32();
    switch (tag >>> 3) {
      case 15:
      message.version = reader.uint32();
      break;
      case 1:
      message.name = reader.string();
      break;
      case 2:
      message.features = decodeTile_Feature(reader, reader.uint32());
      break;
      case 3:
      message.keys = reader.string();
      break;
      case 4:
      message.values = decodeTile_Value(reader, reader.uint32());
      break;
      case 5:
      message.extent = reader.uint32();
      break;
      default:
      reader.skipType(tag & 7);
      break;
    }
  }
  return message;
}

export function encodeLayer(message: Layer, writer: Writer = new Writer()): Writer {
  writer.uint32(120).uint32(message.version);
  writer.uint32(10).string(message.name);
  encodeTile_Feature(message.features, writer.uint32(18).fork()).ldelim();
  writer.uint32(26).string(message.keys);
  encodeTile_Value(message.values, writer.uint32(34).fork()).ldelim();
  writer.uint32(40).uint32(message.extent);
  return writer;
}

export interface Tile_Layer {

  version: number;

  name: string;

  features: Array<Tile_Feature>;

  keys: Array<string>;

  values: Array<Tile_Value>;

  extent: number;

}

export function decodeTile(reader: Reader, length?: number): Tile {
  let end = length === undefined ? reader.len : reader.pos + length;
  const message = {} as Tile;
  while (reader.pos < end) {
    const tag = reader.uint32();
    switch (tag >>> 3) {
      case 3:
      message.layers = decodeTile_Layer(reader, reader.uint32());
      break;
      default:
      reader.skipType(tag & 7);
      break;
    }
  }
  return message;
}

export function encodeTile(message: Tile, writer: Writer = new Writer()): Writer {
  encodeTile_Layer(message.layers, writer.uint32(26).fork()).ldelim();
  return writer;
}

export interface Tile {

  layers: Array<Tile_Layer>;

}
"
`;
